x/staking/keeper/cons_pubkey.go

```diff
@@ -90,3 +132,88 @@ func bytesSliceExists(sliceList [][]byte, targetBytes []byte) bool {
 	}
 	return false
 }
+
+// UpdateAllMaturedConsKeyRotatedKeys udpates all the matured key rotations.
+func (k Keeper) UpdateAllMaturedConsKeyRotatedKeys(ctx sdk.Context, maturedTime time.Time) error {
+	maturedRotatedValAddrs, err := k.GetAllMaturedRotatedKeys(ctx, maturedTime)
+	if err != nil {
+		return err
+	}
+
+	for _, valAddr := range maturedRotatedValAddrs {
+		err := k.deleteConsKeyIndexKey(ctx, valAddr, maturedTime)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deleteConsKeyIndexKey deletes the key which is formed with the given valAddr, time.
+func (k Keeper) deleteConsKeyIndexKey(ctx sdk.Context, valAddr sdk.ValAddress, ts time.Time) error {
+	rng := new(collections.Range[collections.Pair[[]byte, time.Time]]).
+		EndInclusive(collections.Join(valAddr.Bytes(), ts))
+
+	return k.ValidatorConsensusKeyRotationRecordIndexKey.Walk(ctx, rng, func(key collections.Pair[[]byte, time.Time]) (stop bool, err error) {
+		k.ValidatorConsensusKeyRotationRecordIndexKey.Remove(ctx, key)
+		return false, nil
+	})
+}
+
+// GetAllMaturedRotatedKeys returns all matured valaddresses .
+func (k Keeper) GetAllMaturedRotatedKeys(ctx sdk.Context, matureTime time.Time) ([][]byte, error) {
+	valAddrs := [][]byte{}
+
+	// get an iterator for all timeslices from time 0 until the current HeaderInfo time
+	rng := new(collections.Range[time.Time]).EndInclusive(matureTime)
+	iterator, err := k.ValidatorConsensusKeyRotationRecordQueue.Iterate(ctx, rng)
+	if err != nil {
+		return nil, err
+	}
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		value, err := iterator.Value()
+		if err != nil {
+			return nil, err
+		}
+		valAddrs = append(valAddrs, value.Addresses...)
+		key, err := iterator.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		k.ValidatorConsensusKeyRotationRecordQueue.Remove(ctx, key)
+	}
+
+	return valAddrs, nil
+}
+
+// GetBlockConsPubKeyRotationHistory iterator over the rotation history for the given height.
+func (k Keeper) GetBlockConsPubKeyRotationHistory(ctx context.Context) ([]types.ConsPubKeyRotationHistory, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	var historyObjects []types.ConsPubKeyRotationHistory
+
+	iterator, err := k.RotationHistory.Indexes.Block.MatchExact(ctx, uint64(sdkCtx.BlockHeight()))
+	if err != nil {
+		return nil, err
+	}
+	defer iterator.Close()
+
+	keys, err := iterator.PrimaryKeys()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range keys {
+		history, err := k.RotationHistory.Get(ctx, v)
+		if err != nil {
+			return nil, err
+		}
+
+		historyObjects = append(historyObjects, history)
+	}
+
+	return historyObjects, nil
```

Comment:
you could use this: https://github.com/cosmos/cosmos-sdk/blob/main/collections/indexes/helpers.go#L73
```suggestion
	return indexes.CollectValues(ctx, iterator, k.RotationHistory)
```